python/taichi/misc/timing.py

```diff
@@ -0,0 +1,134 @@
+import time, functools
+from colorama import Fore, Back, Style
+
+
+def _m(sec):
+    if sec > 9.9:
+        return f'{Fore.LIGHTRED_EX}{sec:6.1f}{Fore.RESET}s'
+    elif sec > 0.099:
+        return f'{Fore.LIGHTRED_EX}{sec:6.3f}{Fore.RESET}s'
+    else:
+        ms = sec * 1000
+        if ms > 0.099:
+            return f'{Fore.LIGHTMAGENTA_EX}{ms:5.2f}{Fore.RESET}ms'
+        else:
+            ns = ms * 1000
+            return f'{Fore.BLUE}{ns:5.2f}{Fore.RESET}ns'
+
+
+def _c(cnt):
+    return f'{Fore.MAGENTA}{cnt:4d}{Fore.RESET}x'
+
+
+def _n(name):
+    return f'{Fore.MAGENTA}{name}{Fore.RESET}'
+
+
+class RecordStatistics:
+    def __init__(self, profiler, name):
+        self.record = profiler.records[name]
+        self.options = profiler.options[name]
+        self.rec = self.record
+
+        warmup = self.options.get('warmup', 0)
+        self.rec = self.rec[warmup:]
+
+    @property
+    def min(self):
+        return min(self.rec)
+
+    @property
+    def max(self):
+        return max(self.rec)
+
+    @property
+    def total(self):
+        return sum(self.rec)
+
+    @property
+    def avg(self):
+        return sum(self.rec) / len(self.rec)
+
+    @property
+    def count(self):
+        return len(self.rec)
+
+
+class PythonProfiler:
```

Comment:
Maybe we should reuse the C++ `ScopedProfiler` here? The code in this file is mostly a Python rewrite of that C++ class.

Also note that if a timed Python function calls any other timed C++ function, it would be nice to have a hierarchy containing both Python and C++ function calls.

Reply:
> hierarchy

Oh? Is that? Not sure about how `ScopedProfiler` works, I'll dig deeper once I've time.
My concern is that will invocation into pybind11 time cost make its result inaccurate?

Reply:
My another concern is: `ScopedProfiler` is default on and will record profiling data in C++ side. Will this confuse users who only want to test a simple Python side kernel?

Reply:
> My concern is that will invocation into pybind11 time cost make its result inaccurate?

That should be fine. If they really want accuracy directly using `time.time()` would be a better choice.



> My another concern is: `ScopedProfiler` is default on and will record profiling data in C++ side. Will this confuse users who only want to test a simple Python side kernel?

That's a good point. Maybe we can make `ScopeProfiler` off by default and add an option to turn it on?

Reply:
Yes, like `ti.init(cpp_scoped_profiler=True)`, but maybe iapr.